core/src/main/java/org/apache/iceberg/PartitionSpecUpdate.java

```diff
@@ -0,0 +1,214 @@
+    for (PartitionField field : base.spec().fields()) {
+      this.newSpecBuilder.add(field.sourceId(), field.fieldId(), field.name(), field.transform().toString());
+    }
+  }
+
+  @Override
+  public PartitionSpec apply() {
+    Preconditions.checkNotNull(newSpecBuilder, "new partition spec is not set");
+    return newSpecBuilder.build();
+  }
+
+  @Override
+  public UpdatePartitionSpec clear() {
+    this.newSpecBuilder = PartitionSpec.builderFor(base.schema());
+    return this;
+  }
+
+  @Override
+  public UpdatePartitionSpec addIdentityField(String sourceName, String targetName) {
+    newSpecBuilder.identity(sourceName, targetName);
+    return this;
+  }
+
+  @Override
+  public UpdatePartitionSpec addIdentityField(String sourceName) {
+    newSpecBuilder.identity(sourceName);
+    return this;
+  }
+
+  @Override
+  public UpdatePartitionSpec addYearField(String sourceName, String targetName) {
+    newSpecBuilder.year(sourceName, targetName);
+    return this;
+  }
+
+  @Override
+  public UpdatePartitionSpec addYearField(String sourceName) {
+    newSpecBuilder.year(sourceName);
+    return this;
+  }
+
+  @Override
+  public UpdatePartitionSpec addMonthField(String sourceName, String targetName) {
+    newSpecBuilder.month(sourceName, targetName);
+    return this;
+  }
+
+  @Override
+  public UpdatePartitionSpec addMonthField(String sourceName) {
+    newSpecBuilder.month(sourceName);
+    return this;
+  }
+
+  @Override
+  public UpdatePartitionSpec addDayField(String sourceName, String targetName) {
+    newSpecBuilder.day(sourceName, targetName);
+    return this;
+  }
+
+  @Override
+  public UpdatePartitionSpec addDayField(String sourceName) {
+    newSpecBuilder.day(sourceName);
+    return this;
+  }
+
+  @Override
+  public UpdatePartitionSpec addHourField(String sourceName, String targetName) {
+    newSpecBuilder.hour(sourceName, targetName);
+    return this;
+  }
+
+  @Override
+  public UpdatePartitionSpec addHourField(String sourceName) {
+    newSpecBuilder.hour(sourceName);
+    return this;
+  }
+
+  @Override
+  public UpdatePartitionSpec addBucketField(String sourceName, int numBuckets, String targetName) {
+    newSpecBuilder.bucket(sourceName, numBuckets, targetName);
+    return this;
+  }
+
+  @Override
+  public UpdatePartitionSpec addBucketField(String sourceName, int numBuckets) {
+    newSpecBuilder.bucket(sourceName, numBuckets);
+    return this;
+  }
+
+  @Override
+  public UpdatePartitionSpec addTruncateField(String sourceName, int width, String targetName) {
+    newSpecBuilder.truncate(sourceName, width, targetName);
+    return this;
+  }
+
+  @Override
+  public UpdatePartitionSpec addTruncateField(String sourceName, int width) {
+    newSpecBuilder.truncate(sourceName, width);
+    return this;
+  }
+
+  @Override
+  public UpdatePartitionSpec addField(int sourceId, String name, String transform) {
+    newSpecBuilder.add(sourceId, name, transform);
```

Comment:
`PartitionSpecBuilder.add` is not public. It is used to reconstruct specs when deserializing, but should not be exposed to users. Similarly, users should not interact with IDs and transform strings through this API.

Reply:
👌